File: src/main/java/com/soapdogg/lispInterpreter/singleton/TokenizerSingleton.java
package com.soapdogg.lispInterpreter.singleton;

import com.soapdogg.lispInterpreter.tokenizer.LineTokenizer;
import com.soapdogg.lispInterpreter.tokenizer.ScannerToLineTransformer;
import com.soapdogg.lispInterpreter.tokenizer.Tokenizer;
import com.soapdogg.lispInterpreter.tokenizer.WordTokenizer;
import lombok.Getter;

@Getter
public enum TokenizerSingleton {
    INSTANCE;

    private final ScannerToLineTransformer scannerToLineTransformer;
    private final WordTokenizer wordTokenizer;
    private final LineTokenizer lineTokenizer;
    private final Tokenizer tokenizer;

    TokenizerSingleton() {
    scannerToLineTransformer = new ScannerToLineTransformer();
        wordTokenizer = WordTokenizer.newInstance(
            GeneratorSingleton.INSTANCE.getTokenGenerator(),
            DeterminerSingleton.INSTANCE.getNumericTokenValueEndIndexDeterminer(),
            DeterminerSingleton.INSTANCE.getLiteralTokenValueEndIndexDeterminer()
        );
        lineTokenizer = new LineTokenizer(
            wordTokenizer
        );
        tokenizer = new Tokenizer(
            scannerToLineTransformer,
            AsserterSingleton.INSTANCE.getLineFormatAsserter(),
            lineTokenizer
        );
    }
}

File: src/main/java/com/soapdogg/lispInterpreter/singleton/TokenizerSingleton.kt
package com.soapdogg.lispInterpreter.singleton

import com.soapdogg.lispInterpreter.tokenizer.LineTokenizer
import com.soapdogg.lispInterpreter.tokenizer.ScannerToLineTransformer
import com.soapdogg.lispInterpreter.tokenizer.Tokenizer
import com.soapdogg.lispInterpreter.tokenizer.WordTokenizer

enum class TokenizerSingleton {
    INSTANCE;

    val scannerToLineTransformer: ScannerToLineTransformer = ScannerToLineTransformer()
    val wordTokenizer: WordTokenizer = WordTokenizer(
        GeneratorSingleton.INSTANCE.tokenGenerator,
        DeterminerSingleton.INSTANCE.numericTokenValueEndIndexDeterminer,
        DeterminerSingleton.INSTANCE.literalTokenValueEndIndexDeterminer
    )
    val lineTokenizer: LineTokenizer
    val tokenizer: Tokenizer

    init {
        lineTokenizer = LineTokenizer(
            wordTokenizer
        )
        tokenizer = Tokenizer(
            scannerToLineTransformer,
            AsserterSingleton.INSTANCE.lineFormatAsserter,
            lineTokenizer
        )
    }
}